File: node-service/tests/infrastructure/productRepository.test.js
const test = require('node:test');
const assert = require('node:assert');

const ProductRepository = require('../../internal/infrastructure/repository/productRepository');
const DomainError = require('../../internal/domain/errors');
const Product = require('../../internal/domain/product');
const FakeClock = require('../../internal/core/clock/FakeClock');

class StoreStub {
  constructor(overrides = {}) {
    this.listProducts = overrides.listProducts || (async () => []);
    this.getProduct = overrides.getProduct || (async () => null);
    this.createProduct =
      overrides.createProduct ||
      (async (product) => ({ ...product, id: 'abc' }));
    this.updateProduct = overrides.updateProduct || (async () => null);
    this.deleteProduct = overrides.deleteProduct || (async () => true);
    this.health = overrides.health || (async () => {});
  }
}

test('ProductRepository delegates create to the configured store', async () => {
  const clock = new FakeClock(new Date('2024-02-03T04:05:06.000Z'));
  let recordedProduct;
  const repo = new ProductRepository(
    new StoreStub({
      createProduct: async (product) => {
        recordedProduct = product;
        return { ...product, id: 'abc' };
      }
    }),
    clock
  );
  const product = new Product({ name: 'Keyboard', price: 10 });
  const created = await repo.create(product);
  assert.strictEqual(created.id, 'abc');
  assert.deepStrictEqual(recordedProduct.createdAt, clock.now());
  assert.deepStrictEqual(recordedProduct.updatedAt, clock.now());
});

test('ProductRepository wraps store errors as DomainError', async () => {
  const repo = new ProductRepository(
    new StoreStub({
      createProduct: async () => {
        throw new Error('boom');
      }
    }),
    new FakeClock(new Date('2023-01-01T00:00:00.000Z'))
  );

  await assert.rejects(() => repo.create(new Product({ name: 'Mouse', price: 20 })), (err) => {
    assert.ok(err instanceof DomainError);
    assert.strictEqual(err.code, 'internal');
    assert.strictEqual(err.message, 'failed to create product');
    return true;
  });
});

File: go-service/internal/infrastructure/repository/product_repository_test.go
package repository

import (
	"context"
	"errors"
	"testing"
	"time"

	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/require"

	"github.com/example/interview-stack/go-service/internal/domain"
	"github.com/example/interview-stack/go-service/internal/infrastructure/clock"
	"github.com/example/interview-stack/go-service/internal/mocks"
)

func TestProductRepositoryDeleteNotFound(t *testing.T) {
	ctx := context.Background()
	store := mocks.NewProductStoreMock(t)
	store.EXPECT().DeleteProduct(mock.Anything, "missing").Return(false, nil)

	repo := NewProductRepository(store, clock.NewFakeClock(time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)))

	err := repo.DeleteProduct(ctx, "missing")
	require.Error(t, err)
	var domainErr *domain.Error
	require.ErrorAs(t, err, &domainErr)
	require.Equal(t, domain.ErrorCodeNotFound, domainErr.Code)
}

func TestProductRepositoryCreateInternalError(t *testing.T) {
	ctx := context.Background()
	expectedErr := errors.New("boom")
	store := mocks.NewProductStoreMock(t)
	store.EXPECT().
		CreateProduct(mock.Anything, mock.AnythingOfType("domain.Product")).
		Return((*domain.Product)(nil), expectedErr)

	repo := NewProductRepository(store, clock.NewFakeClock(time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)))

	_, err := repo.CreateProduct(ctx, domain.Product{Name: "Keyboard"})
	require.Error(t, err)
	require.ErrorIs(t, err, expectedErr)
}

func TestProductRepositorySetsTimestamps(t *testing.T) {
	ctx := context.Background()
	fakeClock := clock.NewFakeClock(time.Date(2023, 5, 6, 7, 8, 9, 0, time.UTC))
	store := mocks.NewProductStoreMock(t)

	var recorded domain.Product
	store.EXPECT().
		CreateProduct(mock.Anything, mock.MatchedBy(func(product domain.Product) bool {
			recorded = product
			return true
		})).
		Return(&domain.Product{}, nil)

	repo := NewProductRepository(store, fakeClock)

	_, err := repo.CreateProduct(ctx, domain.Product{Name: "Keyboard", Price: 10})
	require.NoError(t, err)
	require.Equal(t, fakeClock.Now(), recorded.CreatedAt)
	require.Equal(t, fakeClock.Now(), recorded.UpdatedAt)
}
